### DISCUSSION TITLE
Expose dtCrowdAgentAnimation

### DISCUSSION THREAD

**Discussion by rob-myers** [isaac-mason/recast-navigation-js#458] [General]
Hey, I would like to  cancel an offMeshConnection traversal.

I can detect if one has started via `agent.state() === 2`, but I cannot currently prevent an agent from following it.

Even if I remove the agent from the crowd and re-add it, the connection is followed.

Can we somehow expose `dtCrowdAgentAnimation` so I can set active to be false? 
> https://github.com/recastnavigation/recastnavigation/blob/455a019e7aef99354ac3020f04c1fe3541aa4d19/DetourCrowd/Source/DetourCrowd.cpp#L1415

**Comment by rob-myers** [comment]
I will try extending:
> https://github.com/isaac-mason/recast-navigation-js/blob/07710dccfbcb8ffaf7202905887b82faddc25093/packages/recast-navigation-wasm/src/Crowd.cpp 

---

**Update**: Exposing `dtCrowdAgentAnimation` requires editing the underlying repo e.g.
```cpp
// packages/recast-navigation-wasm/recastnavigation/DetourCrowd/Source/DetourCrowd.cpp
dtCrowdAgentAnimation* dtCrowd::getAgentAnimation(const int idx)
{
	if (idx < 0 || idx >= m_maxAgents)
		return 0;
	return &m_agentAnims[idx];
}
```

We can follow [DEVELOPMENT.md](https://github.com/isaac-mason/recast-navigation-js/blob/main/DEVELOPMENT.md), change `build.sh` to pull from a forked `recastnavigation`, then run `yarn build` inside `packages/recast-navigation-wasm`. Finally we can hook-up these changes:

```js
// locally by changing tsconfig.json in my repo
"paths": {
  "@recast-navigation/core": ["../recast-navigation-js/packages/recast-navigation-core"],
  "@recast-navigation/generators": ["../recast-navigation-js/packages/recast-navigation-generators"],
  "@recast-navigation/three": ["../recast-navigation-js/packages/recast-navigation-three"],
  "@recast-navigation/wasm": ["../recast-navigation-js/packages/recast-navigation-wasm"]
},

// in production by re-publishing scoped modules e.g.
"@recast-navigation/core": "npm:@rob-myers/recast-navigation__core@0.38.1",
"@recast-navigation/generators": "npm:@rob-myers/recast-navigation__generators@0.38.1",
"@recast-navigation/three": "npm:@rob-myers/recast-navigation__three@0.38.1",
"@recast-navigation/wasm": "npm:@rob-myers/recast-navigation__wasm@0.38.1",
```

I'd be interested if there's another way @isaac-mason ? I'm guessing you don't want to extend the underlying repo, even if offMeshConnections are a bit of a mess.

Btw, it seems pretty easy to "smooth" the offMeshConnection traversal.
> https://github.com/rob-myers/recastnavigation/commit/7e5b66ee299ff1e2a86327c4926acd6536212607

**Comment by isaac-mason** [comment]
Hey @rob-myers!

If you have the appetite to do it, it would be most ideal to upstream improvements to the recastnavigation repo. Although afaik Mikko is not actively involved anymore, there are other contributors maintaining the repo that could help merge improvements to off mesh connections.

Otherwise linking recast-navigation-js to your own fork for your own usage is certainly a way forward. I'd probably prefer not to modify the fork that the recast-navigation-js repo uses though to keep maintenance low though.

**Reply by rob-myers** [reply]
Fair enough. The upstream repo expects developers to extend/modify the C++ code themselves, if necessary, so it could be bit tricky requesting my changes. I'll have a go once I've fixed all the problems I'm working on.

Btw what's the right way to simultaneously publish a patch version of core/generators/three/wasm both in their package.json `version` and `dependencies`? I'm currently doing it manually.

**Reply by isaac-mason** [reply]
> Btw what's the right way to simultaneously publish a patch version of core/generators/three/wasm both in their package.json version and dependencies? I'm currently doing it manually.

I use changesets in the repo, so you can use `yarn run change` to create files describing changes, then `yarn run version` to create version bumps from generated changeset files

**Reply by isaac-mason** [reply]
> Fair enough. The upstream repo expects developers to extend/modify the C++ code themselves, if necessary, so it could be bit tricky requesting my changes. I'll have a go once I've fixed all the problems I'm working on.

That's a fair point, and a common & unfortunate rough edge with trying to make a general wasm library - it's hard to make it highly extensible - it's often easier for serious users to start writing their own opinionated wrapper c++ code etc.

An addition along the lines of Unity's NavMeshAgent.autoTraverseOffMeshLink and NavMeshAgent.CompleteOffMeshLink APIs could be a valuable and general enough additions to recastnavigation. These APIs could let users create any off mesh traversal logic without having to change source code. 

https://docs.unity3d.com/6000.0/Documentation/ScriptReference/AI.NavMeshAgent-autoTraverseOffMeshLink.html
https://docs.unity3d.com/6000.0/Documentation/ScriptReference/AI.NavMeshAgent.CompleteOffMeshLink.html
https://docs.unity3d.com/6000.0/Documentation/ScriptReference/AI.NavMeshAgent-isOnOffMeshLink.html

**Reply by rob-myers** [reply]
> yarn run change

When I bump the patch version of core/generators/three/wasm, this bumps the "version" field in each package.json, but it does not simultaneously bump the dependencies. I would like e.g. `@rob-myers/recast-navigation__three@0.38.4` to have dependency `@recast-navigation/core@0.38.4`.

I'm doing:
```sh
yarn change
yarn changeset version
```

Then manually changing the dependency versions.
I wonder if there's a better way?

**Reply by isaac-mason** [reply]
Sorry for the late reply @rob-myers.

You could try adding `@rob-myers/recast-navigation*` to the "fixed" changesets config: https://github.com/isaac-mason/recast-navigation-js/blob/916ca4f30914739dea143560e1ca40d93bb2222d/.changeset/config.json#L5
